src/m3ninx/index/segment/builder/multi_segments_multi_key_postings_list_iter.go

```diff
@@ -0,0 +1,203 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package builder
+
+import (
+	"bytes"
+
+	"github.com/m3db/m3/src/m3ninx/index/segment"
+	"github.com/m3db/m3/src/m3ninx/postings"
+	"github.com/m3db/m3/src/m3ninx/postings/roaring"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	bitmap "github.com/m3db/pilosa/roaring"
+)
+
+var _ segment.FieldsPostingsListIterator = &multiKeyPostingsListIterator{}
+
+type multiKeyPostingsListIterator struct {
+	err              error
+	firstNext        bool
+	closeIters       []keyIterator
+	iters            []keyIterator
+	currIters        []keyIterator
+	segments         []segment.Segment
+	currPostingsList postings.MutableList
+}
+
+func newMultiKeyPostingsListIterator() *multiKeyPostingsListIterator {
+	b := bitmap.NewBitmapWithDefaultPooling(defaultBitmapContainerPooling)
+	i := &multiKeyPostingsListIterator{
+		currPostingsList: roaring.NewPostingsListFromBitmap(b),
+	}
+	i.reset()
+	return i
+}
+
+func (i *multiKeyPostingsListIterator) reset() {
+	i.firstNext = true
+	i.currPostingsList.Reset()
+
+	for j := range i.closeIters {
+		i.closeIters[j] = nil
+	}
+	i.closeIters = i.closeIters[:0]
+
+	for j := range i.iters {
+		i.iters[j] = nil
+	}
+	i.iters = i.iters[:0]
+
+	for j := range i.currIters {
+		i.currIters[j] = nil
+	}
+	i.currIters = i.currIters[:0]
+
+	for j := range i.segments {
+		i.segments[j] = nil
+	}
+	i.segments = i.segments[:0]
+}
+
+func (i *multiKeyPostingsListIterator) add(iter keyIterator, segment segment.Segment) {
+	i.closeIters = append(i.closeIters, iter)
+	i.iters = append(i.iters, iter)
+	i.segments = append(i.segments, segment)
+	i.tryAddCurr(iter)
+}
+
+func (i *multiKeyPostingsListIterator) Next() bool {
+	if i.err != nil {
+		return false
+	}
+	if len(i.iters) == 0 {
+		return false
+	}
+
+	if i.firstNext {
+		i.firstNext = false
+		return true
+	}
+
+	for _, currIter := range i.currIters {
+		currNext := currIter.Next()
+		if currNext {
+			// Next has a value, forward other matching too
+			continue
+		}
+
+		// Remove iter
+		n := len(i.iters)
+		idx := -1
+		for j, iter := range i.iters {
+			if iter == currIter {
+				idx = j
+				break
+			}
+		}
+		i.iters[idx] = i.iters[n-1]
+		i.iters[n-1] = nil
+		i.iters = i.iters[:n-1]
+	}
+	if len(i.iters) == 0 {
+		return false
+	}
+
+	prevField := i.currIters[0].Current()
+
+	// Re-evaluate current value
+	i.currEvaluate()
+
+	// NB(bodu): Build the postings list for this field if the field has changed.
+	currField := i.currIters[0].Current()
+	if !bytes.Equal(prevField, currField) {
+		i.currPostingsList.Reset()
+		for _, segment := range i.segments {
+			reader, err := segment.Reader()
+			if err != nil {
+				i.err = err
+				return false
+			}
+			pl, err := reader.MatchField(currField)
+			if err != nil {
+				i.err = err
+				return false
+			}
+			i.currPostingsList.Union(pl)
```

Comment:
Instead of unioning on the inner loop (so N times union operations for N segments), I believe we want to keep an array then do a `UnionMany(..)` of all of them so we can do a single union of all the of the segments all at once.

Reply:
We'd also need to keep the readers in an array too, in fact we could just keep a slice of readers and do a tighter inner loop after.
```go
type multiKeyPostingsListIterator struct {
  // ...
  currReaders []segment.Reader
  currFieldPostingsLists []postings.List
}

func (i *multiKeyPostingsListIterator) Next() bool {
  // ...

  i.currReaders = i.currReaders[:0]
  i.currFieldPostingsLists = i.currFieldPostingsLists[:0]

  if !bytes.Equal(prevField, currField) {
    i.currPostingsList.Reset()
    for _, segment := range i.segments {
      reader, err := segment.Reader()
      if err != nil {
        i.err = err
        return false
      }

      pl, err := reader.MatchField(currField)
      if err != nil {
        i.err = err
        return false
      }

      i.currReaders = append(i.currReaders, reader)
      i.currFieldPostingsLists = append(i.currFieldPostingsLists, pl)
    }

    i.currPostingsList.UnionMany(i.currFieldPostingsLists...)

    for _, reader := range i.currReaders {
      if err := reader.Close(); err != nil {
      	i.err = err
      	return false
      }
    }
  }
  // ...
}
```